docs/architecture/adr-069-unordered-account.md

```diff
@@ -0,0 +1,177 @@
+# ADR 070: Un-Ordered Transaction Inclusion
+
+## Changelog
+
+* Dec 4, 2023: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+We propose a way to do replay-attack protection without enforcing the order of transactions, it don't use nonce at all. In this way we can support un-ordered transaction inclusion.
+
+## Context
+
+As of today, the nonce value (account sequence number) prevents replay-attack and ensures the transactions from the same sender are included into blocks and executed in sequential order. However it makes it tricky to send many transactions from the same sender concurrently in a reliable way. IBC relayer and crypto exchanges are typical examples of such use cases.
+
+## Decision
+
+We propose to add a boolean field `unordered` to transaction body to mark "un-ordered" transactions.
+
+Un-ordered transactions will bypass the nonce rules and follow the rules described below instead, in contrary, the default ordered transactions are not impacted by this proposal, they'll follow the nonce rules the same as before.
+
+When an un-ordered transaction are included into block, the transaction hash is recorded in a dictionary, new transactions are checked against this dictionary for duplicates, and to prevent the dictionary grow indefinitly, the transaction must specify `timeout_height` for expiration, so it's safe to removed it from the dictionary after it's expired.
+
+The dictionary can be simply implemented as an in-memory golang map, a preliminary analysis shows that the memory consumption won't be too big, for example `32M = 32 * 1024 * 1024` can support 1024 blocks where each block contains 1024 unordered transactions. For safty, we should limit the range of `timeout_height` to prevent very long expiration, and limit the size of the dictionary.
+
+### Transaction Format
+
+```protobuf
+message TxBody {
+  ...
+
+  boolean unordered = 4; 
+}
+```
+
+### `DedupTxHashManager`
+
+```golang
+// can reduce frequency we check the expiration.
+const ExpireCheckInterval = 1
+
+// DedupTxHashManager contains the tx hash dictionary for duplicates checking,
+// and expire them when block number progresses.
+type DedupTxHashManager struct {
+  // tx hash -> expire block number
+  // for duplicates checking and expiration
+  hashes map[TxHash]uint64
+}
+
+func (dtm *DedupTxHashManager) Contains(hash TxHash) (ok bool) {
+  dtm.mutex.RLock()
+  defer dtm.mutex.RUnlock()
+
+  _, ok = dtm.hashes[hash]
+  return
+}
+
+func (dtm *DedupTxHashManager) Size() int {
+  dtm.mutex.RLock()
+  defer dtm.mutex.RUnlock()
+
+  return len(dtm).hashes
+}
+
+func (dtm *DedupTxHashManager) Add(hash TxHash, expire uint64) (ok bool) {
+  dtm.mutex.Lock()
+  defer dtm.mutex.Unlock()
+
+  dtm.hashes[hash] = expire
+  return
+}
+
+// EndBlock remove expired tx hashes, need to wire in abci cycles.
+func (dtm *DedupTxHashManager) EndBlock(ctx sdk.Context) {
```

Comment:
how will this run in endblock? could we not do a quick check against current time and remove everything behind it? could be a `DeleteUntil` then it checks and removes all previous hashes. This could be done async as well if we wanted

Reply:
> how will this run in endblock?

Wire into baseapp.

Yeah, it can run in background, the end blocker could be a trigger.

It could grab a read lock first and iterate to find out expired hashes, then grab write lock to do the deletion, that should have maximum parallel performance

Reply:
> that sounds good. Since everything under current height wont be touched by the state machine, do we need locks?

we need locks since check tx runs concurrently, and we also added a background purging loop.